File: compiler.js
(function() {

var _ = require('./underscore-min.js');
var util = require('util');

function calcStackSize(prog) {
	return 10;
}

function binaryOp(instruction, operands) {
	if (operands.length != 2) {
		throw new Error(
			'Wrong number of operands to binary operator: ' + operands.length);
	}

	return [
		compile(operands[0]),
		compile(operands[1]),
		instruction,
	].join('\n');
}

builtins = {
	"+": function(args) { return binaryOp("iadd", args); },
	"-": function(args) { return binaryOp("iadd", args); },
	"*": function(args) { return binaryOp("iadd", args); },
	"/": function(args) { return binaryOp("iadd", args); },
	"println": function(args) {
		if (args.length != 1) {
			throw new Error('Wrong number of arguments to println');
		}
		
		return [
			"getstatic java/lang/System/out Ljava/io/PrintStream;",
			compile(args[0]),
			"invokevirtual java/io/PrintStream/println(I)V",
		].join('\n');
	},
};	

function genPrimitive(prim) {
	if (typeof prim === 'number') {
		return 'ldc ' + prim;
	} else {
		throw new Error('Unhandled primitive type');
	}
}

function compile(prog) {
	if (util.isArray(prog)) {
		throw new Error('Invalid list in parse tree');
	}

	if (typeof prog === 'object') {
		if (prog.func in builtins) {
			return builtins[prog.func](prog.args);
		} else {
			throw new Error('Unrecognized symbol: ' + prog.func);
		}
	} else {
		return genPrimitive(prog);
	}
}

function genClass(name, contents) {
	return [
		'.class ' + name,
		'.super java/lang/Object',
		'.method public <init>()V',
		'aload_0',
		'invokenonvirtual java/lang/Object/<init>()V',
		'return',
		'.end method',
		contents,
	].join('\n');
}

function genMain(body) {
	return [
		'.method public static main([Ljava/lang/String;)V',
		'.limit stack ' + calcStackSize(body),
		body,
		'return',
		'.end method',
	].join('\n');
}

console.log(genClass("Main", genMain(compile(
	{func: "println",
	 args: [{func: "+",
			 args: [2, 3]}]}))));
	
})()

File: compiler.py
def get_stack_size(prog):
    return 10


def gen_class(name, contents):
    return '\n'.join([
        '.class {0}'.format(name),
        '.super java/lang/Object',
        '.method public <init>()V',
        'aload_0',
        'invokenonvirtual java/lang/Object/<init>()V',
        'return',
        '.end method',
        contents,
    ])


def gen_main(body):
    return '\n'.join([
        '.method public static main([Ljava/lang/String;)V',
        '.limit stack {0}'.format(get_stack_size(body)),
        body,
        'return',
        '.end method',
    ])


def binary_op(instruction, operands):
    if len(operands) != 2:
        raise RuntimeError('Wrong number of operands for binary op')
    return '\n'.join([
        translate(operands[0]),
        translate(operands[1]),
        instruction,
    ])
        
def translate(prog):
    builtins = {
        '+': lambda args: binary_op('iadd', args),
        '-': lambda args: binary_op('isub', args),
        '*': lambda args: binary_op('imul', args),
        '/': lambda args: binary_op('idiv', args),
        'println': lambda args: '\n'.join([
            "getstatic java/lang/System/out Ljava/io/PrintStream;",
            translate(args[0]),
            "invokevirtual java/io/PrintStream/println(I)V",
        ]),
    }
    
    if isinstance(prog, list):
        sym = prog[0]
        args = prog[1:]

        if sym in builtins:
            return builtins[sym](args)
        else:
            raise RuntimeError('Unrecognized builtin')
    else:
        return 'ldc {0}'.format(prog)


def main():
    print gen_class('Main', gen_main(translate(['println', ['+', 2, 3]])))


if __name__ == '__main__':
    main()
